### ISSUE TITLE
Option to skip files without code

### ISSUE THREAD

**Issue by jayvdb** [coala/coala#2821]
In Python 2, an empty `__init__.py` file is meaningful without any content. Some packages like to leave it completely empty. However this will cause errors with some linters, such as `pydocstyle` and the proposed `CopyrightBear` (`flake8-copyright` has a problematic file size strategy for handling this.).

There may be other languages where a file should be excluded from a linter based on whether the file is empty, or nearly empty. If so, a generic approach should be created for determining emptiness.

**Comment by jayvdb** [comment]
I havent been able to find another language where an empty file can have a meaning.  I have added the Python meaning to https://www.rosettacode.org/wiki/Empty_program#Python

However that page shows many languages consider an empty file to be valid, in which case an empty file might be used in test data for parsers of those languages, and running static analysis tools on the empty files is best avoided.

**Comment by Makman2** [comment]
To exclude files without code just exclude it from the file list.
```
ignore = **/__init__.py
```

As empty files in other languages have no meaning, they should be removed from codebase anyway then.

Imo there's no need to have some special checks (which need to be coded into the core which makes it even more complicated) for file emptiness. If linters have problems with empty files, imo it's their problem (I mean an empty file can happen, it's not some not-happening-at-all-case ;D), and to avoid buggy behaviour inside coala it's possible to ignore specific files like above.

**Comment by adtac** [comment]
I think this should be a bear-specific flag. For instance, the `CopyrightBear` could verify if the file is empty pass it to `flake8-copyright` only if it has content. IMO this is too specific to be a coala wide option.

**Comment by jayvdb** [comment]
"Source file has no code." is one of the simplest possible static analysis of a file. It would belong in `coala`. Specific use of that information might belong in the relevant bear only, but that is not relevant for this issue.

**Comment by Makman2** [comment]
So this is a bear proposal?

Then yes agree, this would be a useful analysis to detect unnecessary empty files in a project.

**Comment by sils** [comment]
LineCountBear could have an option for that and an ignore grep e.g. for
init and stuff

**Comment by jayvdb** [comment]
> So this is a bear proposal?

No, it is support for very basic static analysis. A 'this file contains code' function is very generic, and provides a solid foundation for most other types of analysis to be built on.

Another off-shoot of this would be 'what language is this file', which is relevant as a `.php` file might contain only HTML or even XML, and indeed a `.html` file could be HTML or XML.

**Comment by adtac** [comment]
@jayvdb about the langauge detection thing - see https://gitlab.com/coala/coala-utils/blob/master/coala_utils/Extensions.py

**Comment by Makman2** [comment]
Imo checking the file for emptiness is so trivial that we don't need to have a generic function for that:
```
if file:
    pass
```

I think I'm missing your point @jayvdb  ^^

**Comment by jayvdb** [comment]
If that was the case, https://www.rosettacode.org/wiki/Empty_program would be a very simple wiki page.
It isnt.  And that is only one half of the problem -- it doesnt cover what legally can exist in a file and the file is still empty.  e.g. comments, docstrings, but also xml-preambles and php start and end.

**Comment by Makman2** [comment]
ah you mean actually *code* instead of *content*.
So you want to report that a file is empty for this case:
```python
# Inside x.py there is this comment
```

> "Source file has no code." is one of the simplest possible static analysis of a file.

This is then definitely not the case :)

For example: Just a docstring in a python file is still meaningful, as it's module documentation. You can actually do `help(mypythonfile)` and you would get this docstring^^

But we could definitely try to implement algorithms helping with that^^

**Comment by jayvdb** [comment]
Unassigning to due inactivity

**Comment by jayvdb** [comment]
Detecting that the file has content and no code is still one of the simplest analysis tasks that can be done (completely, correctly), because it only requires determining what syntax can have side effects when compiled/executed, and as the analysis rejects any code blocks, it does not need to know what those code blocks are.

Python docstrings are a special case, as they are actually `__doc__ = ...`.
Indeed python `__init__.py` files are also a very unique case where an empty file is a necessary file (in py2).
